{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bioinformatics with Scikit-Bio\n",
    "+ What is Scikit-Bio?\n",
    "+ It is an open-source, BSD-licensed, python package providing data structures, algorithms, and educational resources for bioinformatics.\n",
    "\n",
    "\n",
    "### Installation\n",
    "+ pip install scikit-bio\n",
    "\n",
    "\n",
    "### Features\n",
    "+ Handling sequences: (skbio.io,skbio.sequence,skbio.alignment)\n",
    "+ Phylogenetics: (skbio.tree)\n",
    "+ Workflow Construction: (skbio.workflow)\n",
    "+ Stats and Calculations : (skbio.stats,skbio.diversity)\n",
    "+ Metadata and Utility functionality (skbio.metadata,skbio.util)\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install scikit-bio\n",
    "\n",
    "# Load Packs\n",
    "import skbio\n",
    "\n",
    "# dir(skbio)\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Working with Sequences \n",
    "#### Sequence Analysis\n",
    "+ DNA and RNA Sequence\n",
    "    - A Adenine\n",
    "    - C Cytosine\n",
    "    - G Guanine\n",
    "    - T Thymine\n",
    "    - U Uracil * RNA\n",
    "    \n",
    "\n",
    "+ ATGATCTCGTAA\n",
    "+ ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sequence\n",
       "---------------\n",
       "Stats:\n",
       "    length: 12\n",
       "---------------\n",
       "0 ATGATCTCGT AA"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from skbio import DNA,RNA,Protein,Sequence\n",
    "\n",
    "# Create a Generic Sequence\n",
    "seq1 = Sequence('ATGATCTCGTAA')\n",
    "seq1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATGATCTCGTAA'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Method 1\n",
    "str(seq1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGATCTCGTAA\n"
     ]
    }
   ],
   "source": [
    "print(seq1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DNA\n",
       "--------------------------\n",
       "Stats:\n",
       "    length: 12\n",
       "    has gaps: False\n",
       "    has degenerates: False\n",
       "    has definites: True\n",
       "    GC-content: 33.33%\n",
       "--------------------------\n",
       "0 ATGATCTCGT AA"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Being More Specific\n",
    "dna_seq = DNA('ATGATCTCGTAA')\n",
    "dna_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Add More Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DNA\n",
       "-----------------------------\n",
       "Metadata:\n",
       "    'description': 'GFP'\n",
       "    'id': 'my_sequence'\n",
       "Positional metadata:\n",
       "    'quality': <dtype: int64>\n",
       "Stats:\n",
       "    length: 8\n",
       "    has gaps: False\n",
       "    has degenerates: False\n",
       "    has definites: True\n",
       "    GC-content: 62.50%\n",
       "-----------------------------\n",
       "0 ACCGGGTA"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d = DNA('ACCGGGTA', metadata = {'id':\"my_sequence\", 'description' : \"GFP\"},\n",
    "       positional_metadata = {'quality' : [22, 25, 22, 18, 23, 25, 25, 25]})\n",
    "\n",
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'-',\n",
       " '.',\n",
       " 'A',\n",
       " 'B',\n",
       " 'C',\n",
       " 'D',\n",
       " 'G',\n",
       " 'H',\n",
       " 'K',\n",
       " 'M',\n",
       " 'N',\n",
       " 'R',\n",
       " 'S',\n",
       " 'T',\n",
       " 'V',\n",
       " 'W',\n",
       " 'Y'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_seq.alphabet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sequence Manipulation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Methods and Attributes\n",
    "# dir(dna_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATGATCTCGTAA'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# To get only the data without the stats and metadata using str/print\n",
    "str(dna_seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGATCTCGTAA\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([b'A', b'T', b'G', b'A', b'T', b'C', b'T', b'C', b'G', b'T', b'A',\n",
       "       b'A'], dtype='|S1')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# To get only the data without the stats and metadata using str/print\n",
    "print(dna_seq)\n",
    "\n",
    "# Convert to numpy array of Nucleotides\n",
    "dna_seq.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DNA\n",
       "--------------------------\n",
       "Stats:\n",
       "    length: 12\n",
       "    has gaps: False\n",
       "    has degenerates: False\n",
       "    has definites: True\n",
       "    GC-content: 33.33%\n",
       "--------------------------\n",
       "0 ATGATCTCGT AA"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get Reverse Complement\n",
    "### GC\n",
    "### AT\n",
    "\n",
    "dna_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DNA\n",
       "--------------------------\n",
       "Stats:\n",
       "    length: 12\n",
       "    has gaps: False\n",
       "    has degenerates: False\n",
       "    has definites: True\n",
       "    GC-content: 33.33%\n",
       "--------------------------\n",
       "0 TACTAGAGCA TT"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get Reverse Complement\n",
    "dna_seq.complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TACTAGAGCATT\n"
     ]
    }
   ],
   "source": [
    "print(dna_seq.complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TTACGAGATCAT'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(dna_seq.complement())[::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_seq.count('G')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_seq.index('T')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DNA\n",
       "--------------------------\n",
       "Stats:\n",
       "    length: 3\n",
       "    has gaps: False\n",
       "    has degenerates: False\n",
       "    has definites: True\n",
       "    GC-content: 33.33%\n",
       "--------------------------\n",
       "0 ATG"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_seq[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGATCTCGTAA\n"
     ]
    }
   ],
   "source": [
    "print(dna_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "slice(0, 3, None)\n",
      "ATG\n"
     ]
    }
   ],
   "source": [
    "# Searched for arbitrary patterns using regular expressions\n",
    "for match in dna_seq.find_with_regex('(ATG+)'):\n",
    "    print(match)\n",
    "    print(str(dna_seq[match]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Find Motifs\n",
    "+ In genetics, a sequence motif is a nucleotide or amino-acid sequence pattern that is widespread and has, or is conjectured to have, a biological significance. For proteins, a sequence motif is distinguished from a structural motif,\n",
    "+ a motif formed by the three-dimensional arrangement of amino acids which may or may not be adjacent.\n",
    "+ Sequence motifs are conserved sequences of similar or identical patterns that may occur within nucleic acids (DNA, RNA) or proteins either\n",
    " - within different molecules produced by the same organism or\n",
    " - within molecules from multiple species of organisms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGATCTCGTAA\n"
     ]
    }
   ],
   "source": [
    "print(dna_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "slice(2, 4, None)\n",
      "GA\n",
      "slice(10, 12, None)\n",
      "AA\n"
     ]
    }
   ],
   "source": [
    "# Find Motif\n",
    "for motif_slice in dna_seq.find_motifs('purine-run', min_length=2):\n",
    "    print(motif_slice)\n",
    "    print(str(dna_seq[motif_slice]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Protein Synthesis\n",
    "+ How cells makes proteins\n",
    "+ Transcription\n",
    "+ Translation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DNA\n",
       "--------------------------\n",
       "Stats:\n",
       "    length: 12\n",
       "    has gaps: False\n",
       "    has degenerates: False\n",
       "    has definites: True\n",
       "    GC-content: 33.33%\n",
       "--------------------------\n",
       "0 ATGATCTCGT AA"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RNA\n",
       "--------------------------\n",
       "Stats:\n",
       "    length: 12\n",
       "    has gaps: False\n",
       "    has degenerates: False\n",
       "    has definites: True\n",
       "    GC-content: 33.33%\n",
       "--------------------------\n",
       "0 AUGAUCUCGU AA"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Transciption : DNA to mRNA\n",
    "dna_seq.transcribe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Protein\n",
       "--------------------------\n",
       "Stats:\n",
       "    length: 4\n",
       "    has gaps: False\n",
       "    has degenerates: False\n",
       "    has definites: True\n",
       "    has stops: True\n",
       "--------------------------\n",
       "0 MIS*"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mRNA = dna_seq.transcribe()\n",
    "\n",
    "# translation mRNA to AA\n",
    "mRNA.translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MIS*\n"
     ]
    }
   ],
   "source": [
    "protein = mRNA.translate()\n",
    "\n",
    "print(protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Protein\n",
       "--------------------------\n",
       "Stats:\n",
       "    length: 6\n",
       "    has gaps: False\n",
       "    has degenerates: False\n",
       "    has definites: True\n",
       "    has stops: False\n",
       "--------------------------\n",
       "0 MITITV"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create A Protein Seq\n",
    "from skbio import Protein\n",
    "protein_a = Protein('MITITV')\n",
    "protein_a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dir(protein_a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sequence Composition\n",
    "+ GC, AT content\n",
    "+ automatically calculates the gc_content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DNA\n",
       "--------------------------\n",
       "Stats:\n",
       "    length: 12\n",
       "    has gaps: False\n",
       "    has degenerates: False\n",
       "    has definites: True\n",
       "    GC-content: 33.33%\n",
       "--------------------------\n",
       "0 ATGATCTCGT AA"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.3333333333333333"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get GC Frequency: Returns a percentage of GC as compared to the whole\n",
    "dna_seq.gc_content()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get GC Frequency: Counts the number of G and C in our Sequence\n",
    "dna_seq.gc_frequency()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.5714285714285714\n",
      "4\n"
     ]
    }
   ],
   "source": [
    "# Difference Between GC_content and GC_frequency\n",
    "print(DNA('ATCGGTC').gc_content())\n",
    "print(DNA('ATCGGTC').gc_frequency())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_seq.count('G')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33.33333333333333"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Custom Fxn  GC Content\n",
    "def gc_content(seq):\n",
    "    return float(seq.count('G') + seq.count('C'))/len(seq) * 100\n",
    "\n",
    "gc_content(dna_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Custom Fxn GC Freq\n",
    "def gc_freq(seq):\n",
    "    return seq.count('G') + seq.count('C')\n",
    "\n",
    "gc_freq(dna_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "66.66666666666666"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# AT Content and AT Freq\n",
    "\n",
    "# Custom Fxn  AT Content\n",
    "def at_content(seq):\n",
    "    return float(seq.count('A') + seq.count('T'))/len(seq) * 100\n",
    "\n",
    "at_content(dna_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Custom Fxn  AT Freq\n",
    "def at_freq(seq):\n",
    "    return seq.count('A') + seq.count('T')\n",
    "\n",
    "at_freq(dna_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'A': 4, 'C': 2, 'G': 2, 'T': 4}"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get Frequencies of Each Nucleotide\n",
    "dna_seq.frequencies()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'A': 4, 'T': 4, 'G': 2, 'C': 2})"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Custom Fxn Method 1\n",
    "from collections import Counter\n",
    "Counter(str(dna_seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BarContainer object of 4 artists>"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAD4CAYAAAD8Zh1EAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAO00lEQVR4nO3df6hfd33H8edrMU6HQmC52JAfjRsZwxZau7vY0oFdUWhrWZT1j2TMQhkLLRUUBNf5R53/uX/caCO9hFm0zFkEXQmaTgq12MJSe5Ol0TQ6Mn/Qu2b0WjE1tOgi7/1xT+Xy7ffme76539tv7mfPBxzu+fH+nvPmkLxycu7nfE+qCknS+vdb025AkjQZBrokNcJAl6RGGOiS1AgDXZIa8aZpHXjz5s21c+fOaR1ektalo0eP/rSqZoZtm1qg79y5k/n5+WkdXpLWpSQ/WWmbt1wkqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSI3oHepINSf4jydeHbEuS+5KcTnIiyTWTbVOSNMo4V+gfBU6tsO1mYFc37QceWGVfkqQx9Qr0JNuADwD/tELJHuChWnIE2JRky4R6lCT10PdJ0X8EPgG8fYXtW4Hnly0vdOvOLC9Ksp+lK3h27NgxVqPL7bznGxf92Rb8+DMfmHYL0qr4d3ht/g6PvEJPcivwYlUdvVDZkHWvexVSVR2sqtmqmp2ZGfpVBJKki9Tnlsv1wJ8l+THwMHBjkn8eqFkAti9b3ga8MJEOJUm9jAz0qvrbqtpWVTuBvcDjVfWXA2WHgNu70S7XAmer6szgviRJa+eiv20xyZ0AVTUHHAZuAU4DrwB3TKQ7SVJvYwV6VT0BPNHNzy1bX8Ddk2xMkjQenxSVpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDWiz0ui35LkO0meTXIyyaeH1NyQ5GyS491079q0K0laSZ83Fv0SuLGqziXZCDyV5NGqOjJQ92RV3Tr5FiVJfYwM9O71cue6xY3dVGvZlCRpfL3uoSfZkOQ48CLwWFU9PaTsuu62zKNJrphol5KkkXoFelX9uqquBrYBu5NcOVByDLi8qq4C7gceGbafJPuTzCeZX1xcXE3fkqQBY41yqaqfA08ANw2sf7mqznXzh4GNSTYP+fzBqpqtqtmZmZmL71qS9Dp9RrnMJNnUzb8VeB/w/YGay5Kkm9/d7felybcrSVpJn1EuW4AvJtnAUlB/paq+nuROgKqaA24D7kpyHngV2Nv9MlWS9AbpM8rlBPDuIevnls0fAA5MtjVJ0jh8UlSSGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIa0eedom9J8p0kzyY5meTTQ2qS5L4kp5OcSHLN2rQrSVpJn3eK/hK4sarOJdkIPJXk0ao6sqzmZmBXN70HeKD7KUl6g4y8Qq8l57rFjd00+ALoPcBDXe0RYFOSLZNtVZJ0Ib3uoSfZkOQ48CLwWFU9PVCyFXh+2fJCt25wP/uTzCeZX1xcvNieJUlD9Ar0qvp1VV0NbAN2J7lyoCTDPjZkPweraraqZmdmZsbvVpK0orFGuVTVz4EngJsGNi0A25ctbwNeWFVnkqSx9BnlMpNkUzf/VuB9wPcHyg4Bt3ejXa4FzlbVmYl3K0laUZ9RLluALybZwNI/AF+pqq8nuROgquaAw8AtwGngFeCONepXkrSCkYFeVSeAdw9ZP7dsvoC7J9uaJGkcPikqSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjejzTtHtSb6V5FSSk0k+OqTmhiRnkxzvpnvXpl1J0kr6vFP0PPDxqjqW5O3A0SSPVdVzA3VPVtWtk29RktTHyCv0qjpTVce6+V8Ap4Cta92YJGk8Y91DT7KTpRdGPz1k83VJnk3yaJIrVvj8/iTzSeYXFxfHblaStLLegZ7kbcBXgY9V1csDm48Bl1fVVcD9wCPD9lFVB6tqtqpmZ2ZmLrZnSdIQvQI9yUaWwvxLVfW1we1V9XJVnevmDwMbk2yeaKeSpAvqM8olwOeBU1X12RVqLuvqSLK72+9Lk2xUknRhfUa5XA98GPhukuPduk8COwCqag64DbgryXngVWBvVdUa9CtJWsHIQK+qp4CMqDkAHJhUU5Kk8fmkqCQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDWizztFtyf5VpJTSU4m+eiQmiS5L8npJCeSXLM27UqSVtLnnaLngY9X1bEkbweOJnmsqp5bVnMzsKub3gM80P2UJL1BRl6hV9WZqjrWzf8COAVsHSjbAzxUS44Am5JsmXi3kqQV9blC/40kO4F3A08PbNoKPL9seaFbd2bg8/uB/QA7duwYr1NN1M57vjHtFqbqx5/5wKo+7/lb3fnT2uj9S9EkbwO+Cnysql4e3DzkI/W6FVUHq2q2qmZnZmbG61SSdEG9Aj3JRpbC/EtV9bUhJQvA9mXL24AXVt+eJKmvPqNcAnweOFVVn12h7BBwezfa5VrgbFWdWaFWkrQG+txDvx74MPDdJMe7dZ8EdgBU1RxwGLgFOA28Atwx+VYlSRcyMtCr6imG3yNfXlPA3ZNqSpI0Pp8UlaRGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEb0eafog0leTPK9FbbfkORskuPddO/k25QkjdLnnaJfAA4AD12g5smqunUiHUmSLsrIK/Sq+jbwszegF0nSKkzqHvp1SZ5N8miSK1YqSrI/yXyS+cXFxQkdWpIEkwn0Y8DlVXUVcD/wyEqFVXWwqmaranZmZmYCh5YkvWbVgV5VL1fVuW7+MLAxyeZVdyZJGsuqAz3JZUnSze/u9vnSavcrSRrPyFEuSb4M3ABsTrIAfArYCFBVc8BtwF1JzgOvAnurqtasY0nSUCMDvar2jdh+gKVhjZKkKfJJUUlqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWrEyEBP8mCSF5N8b4XtSXJfktNJTiS5ZvJtSpJG6XOF/gXgpgtsvxnY1U37gQdW35YkaVwjA72qvg387AIle4CHaskRYFOSLZNqUJLUzyTuoW8Fnl+2vNCte50k+5PMJ5lfXFycwKElSa+ZRKBnyLoaVlhVB6tqtqpmZ2ZmJnBoSdJrJhHoC8D2ZcvbgBcmsF9J0hgmEeiHgNu70S7XAmer6swE9itJGsObRhUk+TJwA7A5yQLwKWAjQFXNAYeBW4DTwCvAHWvVrCRpZSMDvar2jdhewN0T60iSdFF8UlSSGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIa0SvQk9yU5AdJTie5Z8j2G5KcTXK8m+6dfKuSpAvp807RDcDngPcDC8AzSQ5V1XMDpU9W1a1r0KMkqYc+V+i7gdNV9cOq+hXwMLBnbduSJI2rT6BvBZ5ftrzQrRt0XZJnkzya5IphO0qyP8l8kvnFxcWLaFeStJI+gZ4h62pg+RhweVVdBdwPPDJsR1V1sKpmq2p2ZmZmvE4lSRfUJ9AXgO3LlrcBLywvqKqXq+pcN38Y2Jhk88S6lCSN1CfQnwF2JXlnkjcDe4FDywuSXJYk3fzubr8vTbpZSdLKRo5yqarzST4CfBPYADxYVSeT3NltnwNuA+5Kch54FdhbVYO3ZSRJa2hkoMNvbqMcHlg3t2z+AHBgsq1Jksbhk6KS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUiF6BnuSmJD9IcjrJPUO2J8l93fYTSa6ZfKuSpAsZGehJNgCfA24G3gXsS/KugbKbgV3dtB94YMJ9SpJG6HOFvhs4XVU/rKpfAQ8DewZq9gAP1ZIjwKYkWybcqyTpAvq8JHor8Pyy5QXgPT1qtgJnlhcl2c/SFTzAuSQ/GKvbS8dm4KfTOnj+flpHnijP4ep4/lZnPZ+/y1fa0CfQM2RdXUQNVXUQONjjmJe0JPNVNTvtPtYzz+HqeP5Wp9Xz1+eWywKwfdnyNuCFi6iRJK2hPoH+DLAryTuTvBnYCxwaqDkE3N6NdrkWOFtVZwZ3JElaOyNvuVTV+SQfAb4JbAAerKqTSe7sts8Bh4FbgNPAK8Ada9fyJWHd3za6BHgOV8fztzpNnr9Uve5WtyRpHfJJUUlqhIEuSY0w0MeU5ENJKskfTruX9SjJZUkeTvJfSZ5LcjjJH0y7r/UgyTuS/EuSHyY5muTfk3xo2n2tF0l+N8nxbvqfJP+9bPnN0+5vEgz08e0DnmJptI/GkCTAvwJPVNXvV9W7gE8C75huZ5e+7tw9Any7qn6vqv6IpT+D26bb2fpRVS9V1dVVdTUwB/zDa8vdU/DrnoE+hiRvA64H/goD/WL8KfC/3cgoAKrqeFU9OcWe1osbgV8NnLufVNX9U+xJlxgDfTwfBP6tqv4T+JnfKjm2K4Gj025inboCODbtJnRpM9DHs4+lLyej+7lvir3o/7Ekn0vybJJnpt2LLh19vstFLP1ChaX/9l6ZpFh6yKqSfKIczN/XSeC2aTexTp0E/vy1haq6O8lmYH56LelS4xV6f7ex9BXBl1fVzqraDvwI+JMp97WePA78dpK/fm1Fkj9O8t4p9rRePA68Jcldy9b9zrSa0aXJQO9vH0sjNJb7KvAXU+hlXer+J/Mh4P3dsMWTwN/hF7mN1J27DwLvTfKjJN8Bvgj8zXQ706XER/8lqRFeoUtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1Ij/A3N77kR3zz/sAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Plot of N Freq\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "plt.bar(dna_seq.frequencies().keys(),dna_seq.frequencies().values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'C': 0.16666666666666666, 'A': 0.3333333333333333}"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check for The Freq of A Specific N\n",
    "dna_seq.frequencies(chars={'A','C'},relative=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Kmers\n",
    "+ k-mers are subsequences of length contained within a biological sequence. \n",
    "+ More generally, a sequence of length \n",
    " - L will have `L-k+1` k-mers and `n^K` total possible k-mers, \n",
    " - where n is number of possible monomers\n",
    " \n",
    "#### Similar to ngrams in NLP\n",
    " \n",
    "##### Usefulness\n",
    "+  assemble DNA sequences,\n",
    "+ improve heterologous gene expression,\n",
    "+ identify species in metagenomic samples\n",
    "+ create attenuated vaccines\n",
    "\n",
    "![](./images/kmerimage.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGATCTCGTAA\n"
     ]
    }
   ],
   "source": [
    "print(dna_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ATG': 1,\n",
       " 'TGA': 1,\n",
       " 'GAT': 1,\n",
       " 'ATC': 1,\n",
       " 'TCT': 1,\n",
       " 'CTC': 1,\n",
       " 'TCG': 1,\n",
       " 'CGT': 1,\n",
       " 'GTA': 1,\n",
       " 'TAA': 1}"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find Kmers\n",
    "dna_seq.kmer_frequencies(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ATGA': 1,\n",
       " 'TGAT': 1,\n",
       " 'GATC': 1,\n",
       " 'ATCT': 1,\n",
       " 'TCTC': 1,\n",
       " 'CTCG': 1,\n",
       " 'TCGT': 1,\n",
       " 'CGTA': 1,\n",
       " 'GTAA': 1}"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find Kmers\n",
    "dna_seq.kmer_frequencies(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ATG': 1,\n",
       " 'TGA': 1,\n",
       " 'GAT': 1,\n",
       " 'ATC': 1,\n",
       " 'TCT': 1,\n",
       " 'CTC': 1,\n",
       " 'TCG': 1,\n",
       " 'CGT': 1,\n",
       " 'GTA': 1,\n",
       " 'TAA': 1}"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Custom Fxn KMer\n",
    "# L - k+1\n",
    "def count_kmers(seq,k=3):\n",
    "    # Empty Dictionary\n",
    "    counts = {}\n",
    "    num_kmers = len(seq) - k + 1\n",
    "    for i in range(num_kmers):\n",
    "        # Slicing the seq to get the kmer\n",
    "        kmer  = seq[i:i+k]\n",
    "        if kmer not in counts:\n",
    "            counts[kmer] = 0\n",
    "        # increment\n",
    "        #counts[kmer]  = counts[kmer] + 1\n",
    "        counts[kmer] += 1\n",
    "    return counts\n",
    "\n",
    "count_kmers(str(dna_seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ATGA': 1,\n",
       " 'TGAT': 1,\n",
       " 'GATC': 1,\n",
       " 'ATCT': 1,\n",
       " 'TCTC': 1,\n",
       " 'CTCG': 1,\n",
       " 'TCGT': 1,\n",
       " 'CGTA': 1,\n",
       " 'GTAA': 1}"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_kmers(str(dna_seq),k=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATG\n",
      "TGA\n",
      "GAT\n",
      "ATC\n",
      "TCT\n",
      "CTC\n",
      "TCG\n",
      "CGT\n",
      "GTA\n",
      "TAA\n"
     ]
    }
   ],
   "source": [
    "# iterating through a kmer\n",
    "for k in dna_seq.iter_kmers(3,overlap=True):\n",
    "    print(k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Kmer Freq\n",
    "dna_kfreq = dna_seq.kmer_frequencies(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_seq1 = DNA('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAA')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BarContainer object of 10 artists>"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAD4CAYAAAD8Zh1EAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAQnUlEQVR4nO3df7BcdXnH8ffHREAriIVb1PwwaRu10YqjEZ2qFWvVhFqRlqnEVgojpqlg+0dtodWpzmgV63SmWpBMain6R804I2rEVGasvytokoqYYKFXUIixNZQOOlKL0ad/7ImzLHvvbm7ODeXb92vmTvac8919nnPu2c+ee3b3JFWFJOnB7yEPdAOSpH4Y6JLUCANdkhphoEtSIwx0SWrE0geq8Mknn1yrVq16oMpL0oPS7t2776yqmXHLHrBAX7VqFbt27XqgykvSg1KSb861zFMuktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqRETAz3JlUm+k2TPHMuT5F1JZpPcmORp/bcpSZpkmiP0q4D18yzfAKzpfjYBVxx5W5KkwzUx0Kvqs8Bd8ww5E3hfDVwPnJjkMX01KEmaTh/fFF0G3DE0va+b9+3RgUk2MTiKZ+XKlQsuuOqSjy34vtP6xqW/9n+u9tGob+2jX3u++tb+/1X7SPXxpmjGzBv73yBV1daqWldV62Zmxl6KQJK0QH0E+j5gxdD0cmB/D48rSToMfQT6duDc7tMuzwLurqr7nW6RJC2uiefQk7wfOB04Ock+4I3AQwGqaguwAzgDmAXuAc5frGYlSXObGOhVtXHC8gIu7K0jSdKC+E1RSWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiOmCvQk65PcnGQ2ySVjlj8yyUeTfCXJ3iTn99+qJGk+EwM9yRLgcmADsBbYmGTtyLALgZuq6lTgdOCvkhzTc6+SpHlMc4R+GjBbVbdW1b3ANuDMkTEFHJ8kwCOAu4CDvXYqSZrXNIG+DLhjaHpfN2/YZcAvAPuBrwJ/WFU/Hn2gJJuS7Eqy68CBAwtsWZI0zjSBnjHzamT6xcANwGOBpwKXJTnhfneq2lpV66pq3czMzGE3K0ma2zSBvg9YMTS9nMGR+LDzgatrYBa4DXhiPy1KkqYxTaDvBNYkWd290XkOsH1kzO3ACwCSnAI8Abi1z0YlSfNbOmlAVR1MchFwLbAEuLKq9ibZ3C3fArwZuCrJVxmcorm4qu5cxL4lSSMmBjpAVe0AdozM2zJ0ez/won5bkyQdDr8pKkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWrEVIGeZH2Sm5PMJrlkjjGnJ7khyd4kn+m3TUnSJEsnDUiyBLgceCGwD9iZZHtV3TQ05kTg3cD6qro9yc8sVsOSpPGmOUI/DZitqlur6l5gG3DmyJhXAFdX1e0AVfWdftuUJE0yTaAvA+4Ymt7XzRv2eOBRST6dZHeSc/tqUJI0nYmnXICMmVdjHufpwAuAhwHXJbm+qm65zwMlm4BNACtXrjz8biVJc5rmCH0fsGJoejmwf8yYj1fV96vqTuCzwKmjD1RVW6tqXVWtm5mZWWjPkqQxpgn0ncCaJKuTHAOcA2wfGfMR4LlJliZ5OPBM4Gv9tipJms/EUy5VdTDJRcC1wBLgyqram2Rzt3xLVX0tyceBG4EfA++pqj2L2bgk6b6mOYdOVe0AdozM2zIy/Q7gHf21Jkk6HH5TVJIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGjFVoCdZn+TmJLNJLpln3DOS/CjJ2f21KEmaxsRAT7IEuBzYAKwFNiZZO8e4twPX9t2kJGmyaY7QTwNmq+rWqroX2AacOWbca4EPAt/psT9J0pSmCfRlwB1D0/u6eT+RZBlwFrBlvgdKsinJriS7Dhw4cLi9SpLmMU2gZ8y8Gpn+a+DiqvrRfA9UVVural1VrZuZmZm2R0nSFJZOMWYfsGJoejmwf2TMOmBbEoCTgTOSHKyqD/fSpSRpomkCfSewJslq4FvAOcArhgdU1epDt5NcBVxjmEvS0TUx0KvqYJKLGHx6ZQlwZVXtTbK5Wz7veXNJ0tExzRE6VbUD2DEyb2yQV9V5R96WJOlw+U1RSWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiOmCvQk65PcnGQ2ySVjlv92khu7ny8kObX/ViVJ85kY6EmWAJcDG4C1wMYka0eG3QY8r6qeArwZ2Np3o5Kk+U1zhH4aMFtVt1bVvcA24MzhAVX1har6r27yemB5v21KkiaZJtCXAXcMTe/r5s3lVcA/jluQZFOSXUl2HThwYPouJUkTTRPoGTOvxg5Mns8g0C8et7yqtlbVuqpaNzMzM32XkqSJlk4xZh+wYmh6ObB/dFCSpwDvATZU1X/2054kaVrTHKHvBNYkWZ3kGOAcYPvwgCQrgauBV1bVLf23KUmaZOIRelUdTHIRcC2wBLiyqvYm2dwt3wL8OXAS8O4kAAerat3itS1JGjXNKReqagewY2TelqHbFwAX9NuaJOlw+E1RSWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqxFSBnmR9kpuTzCa5ZMzyJHlXt/zGJE/rv1VJ0nwmBnqSJcDlwAZgLbAxydqRYRuANd3PJuCKnvuUJE0wzRH6acBsVd1aVfcC24AzR8acCbyvBq4HTkzymJ57lSTNI1U1/4DkbGB9VV3QTb8SeGZVXTQ05hrg0qr6fDf9T8DFVbVr5LE2MTiCB3gCcHNfKzKFk4E7j2I9a1vb2tZeDI+rqplxC5ZOceeMmTf6KjDNGKpqK7B1ipq9S7KrqtZZ29rWtnYrtUdNc8plH7BiaHo5sH8BYyRJi2iaQN8JrEmyOskxwDnA9pEx24Fzu0+7PAu4u6q+3XOvkqR5TDzlUlUHk1wEXAssAa6sqr1JNnfLtwA7gDOAWeAe4PzFa3nBHpBTPda2trWtfbRMfFNUkvTg4DdFJakRBrokNaKZQE9yVpJK8sQkX0xyQ5Lbkxzobt+QZFWSRyS5IsnXk3w5ye4kr15AvZOGHvffk3xraPqUJP+Q5Nbu8a9LctbI/d/Z3eeIfgeTag3XSfKLQz3eleS27vYnjqD+tNv93CR7kuxNclOS1x1Bzfm2/Z8k+deu1le6uh/qls0muXto7C8tsP6jk2zr9qGbknwqyT3jtmuShya5NMm/dT19KcmGxV7n7j5Lk7y1q31o7OsXss5zrPeOJI9PsibJNd383d32+OU+t/sc+/ne7vFuSvLfQ49/dnefjyS5bqHr2z3GpOf5D5P83pj7LU1yZ5K3HUn9w1ZVTfwAHwA+B7xpaN55wGUj47YBbwUe0k3PMPgS1JHUfhPwuu52gOuAzUPLHwe8dmj6IcDtwPXA6UdQd95a89UBrgLOPhrbncGlIf4FeGw3fRzw6p5+78PbfjODN+9P6KYfCfzu0NjTgWuOsN64bf5U4LnjtitwKfBe4Nhu+hTgt47GOne1rwKO66aPH/499bXewC3AS4fmPxk4r6/tPsV+vgrYM3KfE4E7gK8Bq/ve17rp13T7/qfHjD0D+Gfg63TvVR6NnyaO0JM8Ang28CoGH6uca9zPMbiUwRuq6scAVXWgqt7eYzu/Atxbg0//0NX4ZlX9zdCY5wN7GFzzZuMi1uqrzljTbnfgTxk8EfZ3Pf6gqv62736APwNeU1Xf7ercXVXv7bnG84EfjmzzG6rqc6MDkzwceDWD4Pmfbux/VNUHeuxn7DqP1P5Bt+x7VfWmBdYZu97A44Hrqmr70Pw9VXXVAuuMM81zatRvAh9lcAA33755JDYCfwQsT7JszLJ3MjigetYi1b+fJgIdeBnw8aq6Bbgrc1/t8UnAVw6F+SJ5EoOj0flsBN4PfAh4SZKHLlKtvurMZdrt/mRgd8+17yPJ8cDxVfX1xazD4a3LzwO3Hwrbvk1Y50O1v9dTubnWe5r9/UgtpMahff/9LM7BzArg0VX1JQZ/pb58aNnDgBcA1yxW/bm0EugbGbwS0/071QZM8vruXNiifas1yeXdec2d3fQxDP4c+3D3RP8i8KK+ay1mnSEL2u6LJIy53ETjpl7nJOd3+/odXRgtTkOD8+Z7kly9iDXu85was/wUBi9on+8ONg4meXLPbZzDIMjh/vv+S4BPVdU9wAeBszK4au2ie9AHepKTGPxJ9p4k3wD+GHh5knHXl7kJODXdG5FV9RdV9VTghB5b2gv85Ei1qi5k8Gp96GI66xmc5/xq1+9zWHgQzlerzzr3c5jbfS/w9L5qj9O9aH0/yc8uZh0Ob11mgZXdkXTvJqzzfWpX1d93+/rdDL4geLjmWu/RffAsBu+h/PQCasxXe77n1KiXA48Cbuv2zVX0f9plI3Be9/jbGeTKmqFlv9ot2w2cxOCU1aJ70Ac6cDaDS/c+rqpWVdUK4DYGAXYfVTUL7ALecugVM8lxjL+42EJ9Ejguye8PzXv40O2NwAVdr6uA1cCLunOefdbqs844U2934G3AXyZ5NECSY5P8QU99jNa5PMkJXZ0TMrjCZ58+CRyboU9GJXlGkueNDuyO0P4OeFf3FxNJHpPkd3rsZ+w6D9W+rNvHD/3fBscssM7Y9WbwwvHsJC8dGtvXPjZce77n1KiNDK4Qe2jffzo9BnqSJwA/VVXLhmq8DTin+z08B1g5tOxCjtJfry0E+kYG54iHfRB4xRzjL2DwijmbZDfwCeDivpqpwVvcLwOel8HH177E4FMOF3dh+mLgY0Pjvw98Hvj1Hmu9sc86c5h6u1fVDgb/SconkuxlcNQyzZU+D9cVwKeAnUn2AJ9hcCmK3nTb/CzghRl8TG8vg08/zHXa7g3AAeCmrqcPd9N9mW+dXw98G9iT5MsMPpHx3nl6ndOE9X4JsLn7SOF1DNb5LUe0VvevPfY5NTo2ySpgJYNPdh26/23Ad5M8s6eW5tr3NwK/AXzy0JvgnY8AL01ybE/15+RX/yWpES0coUuSMNAlqRkGuiQ1wkCXpEYY6JLUCANdkhphoEtSI/4X2fpFAmBpvU4AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot of Kmers Freq\n",
    "plt.bar(dna_kfreq.keys(),dna_kfreq.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DNA\n",
       "--------------------------------------------------------------------\n",
       "Stats:\n",
       "    length: 70\n",
       "    has gaps: False\n",
       "    has degenerates: False\n",
       "    has definites: True\n",
       "    GC-content: 35.71%\n",
       "--------------------------------------------------------------------\n",
       "0  ATTAAAGGTT TATACCTTCC CAGGTAACAA ACCAACCAAC TTTCGATCTC TTGTAGATCT\n",
       "60 GTTCTCTAAA"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_seq1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Kmer Freq\n",
    "dna1_kfreq = dna_seq1.kmer_frequencies(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BarContainer object of 37 artists>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABAgAAAFlCAYAAABmy9o5AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAazklEQVR4nO3dfbBtZ10f8O+PXCCgiVhyGyBwc6jy2mABr7ESRaAKgSBCjZprUeLbLZVSbMvQWJjxdZqMZTrikHbMoLzYAjoD0Q5BVAoRsEggmEBeeA2XVy0iYwNitdCnf6x1defes/dea5299zn35POZOXPP3nvt52U/z1r72d+79jrVWgsAAABw53aX3W4AAAAAsPsEBAAAAICAAAAAABAQAAAAABEQAAAAABEQAAAAAEkOrKPQs846q21tba2jaAAAAGCi66+//nOttYPbPbaWgGBrayvvec971lE0AAAAMFFVfXzeY75iAAAAAAgIAAAAAAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAkOTAkI2q6liSLyT5SpIvt9YOr7NRAAAAwGYNCgh6j2+tfW5tLQEAAAB2ja8YAAAAAIPPIGhJfq+qWpJfaa1ddeIGVXU0ydEkOXTo0OpaCAAA7Dlbl12z0vKOXXHRSssDxht6BsEFrbVHJ3lykudU1WNP3KC1dlVr7XBr7fDBgwdX2kgAAABgvQYFBK21z/T/fjbJ1UnOX2ejAAAAgM1aGhBU1VdV1RnHf0/yxCQ3rbthAAAAwOYMuQbB2Umurqrj27+6tfamtbYKAAAA2KilAUFr7bYk/2gDbQEAAAB2iT9zCAAAAAgIAAAAAAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAEAEBAAAAkBEBQVWdVlV/XFVvWGeDAAAAgM0bcwbB85Lcuq6GAAAAALtnUEBQVfdPclGSl623OQAAAMBuODBwu19K8oIkZ8zboKqOJjmaJIcOHdp5ywAA9oCty65ZaXnHrrhopeUBwKosPYOgqp6a5LOttesXbddau6q1dri1dvjgwYMrayAAAACwfkO+YnBBkqdV1bEkr03yhKr6r2ttFQAAALBRSwOC1tpPtdbu31rbSnJJkre01p659pYBAAAAGzPmrxgAAAAA+9TQixQmSVpr1ya5di0tAQAAAHaNMwgAAAAAAQEAAAAgIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAiIAAAAAAyICCoqtOr6rqqurGqbq6qn91EwwAAAIDNOTBgm79O8oTW2her6q5J3lFVv9Na+6M1tw0AAADYkKUBQWutJflif/Ou/U9bZ6MAAACAzRpyBkGq6rQk1yf5+iRXttbetc02R5McTZJDhw6tso2MsHXZNSst79gVF+1KHcDJ9su+t8p+OEbtjNfqzsV4A7DMoIsUtta+0lp7ZJL7Jzm/qs7bZpurWmuHW2uHDx48uOp2AgAAAGs06q8YtNb+Ism1SS5cS2sAAACAXTHkrxgcrKp79b/fI8l3JPnAuhsGAAAAbM6QaxDcN8kr++sQ3CXJb7bW3rDeZgEAAACbNOSvGLwvyaM20BYAAABgl4y6BgEAAACwPwkIAAAAAAEBAAAAICAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAIiAAAAAAMiAgqKoHVNVbq+rWqrq5qp63iYYBAAAAm3NgwDZfTvJvW2vvraozklxfVb/fWrtlzW0DAAAANmTpGQSttT9prb23//0LSW5Ncs66GwYAAABszqhrEFTVVpJHJXnXOhoDAAAA7I4hXzFIklTVVyd5XZKfbK3dvs3jR5McTZJDhw6trIH7ydZl16y0vGNXXLTS8u5MNjEW6hheB8BUjlGcirx/D7fKfuyHPiSOU6zXoDMIququ6cKB/9Zae/1227TWrmqtHW6tHT548OAq2wgAAACs2ZC/YlBJfjXJra21/7T+JgEAAACbNuQMgguS/GCSJ1TVDf3PU9bcLgAAAGCDll6DoLX2jiS1gbYAAAAAu2TUXzEAAAAA9icBAQAAACAgAAAAAAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAQEAAAAQAYEBFX1a1X12aq6aRMNAgAAADZvyBkEr0hy4ZrbAQAAAOyipQFBa+1tST6/gbYAAAAAu8Q1CAAAAIAcWFVBVXU0ydEkOXTo0KqK3Zity65ZaXnHrrhopeXdmRiLO59Vjvl2472JObVf5u1+6ce6mVPD7Zd+rNt+eZ1OtX1jP+97DLNfxvtU2/d2q479Mt7rtLIzCFprV7XWDrfWDh88eHBVxQIAAAAb4CsGAAAAwKA/c/iaJO9M8pCq+lRV/ej6mwUAAABs0tJrELTWjmyiIQAAAMDu8RUDAAAAQEAAAAAACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACADAwIqurCqvpgVX2kqi5bd6MAAACAzVoaEFTVaUmuTPLkJA9PcqSqHr7uhgEAAACbM+QMgvOTfKS1dltr7W+SvDbJd6+3WQAAAMAmDQkIzknyyZnbn+rvAwAAAPaJaq0t3qDqe5M8qbX2Y/3tH0xyfmvtuSdsdzTJ0f7mQ5J8cPXN3RPOSvI5deyJOvZDH9Sxd8pXx52vjv3QB3XsnfLVsbfq2A99UMfeqmM/9EEde6f83XZua+3gdg8cGPDkTyV5wMzt+yf5zIkbtdauSnLVpOadQqrqPa21w+rY/Tr2Qx/UsXfKV8edr4790Ad17J3y1bG36tgPfVDH3qpjP/RBHXun/L1syFcM3p3kQVX1wKq6W5JLkvz39TYLAAAA2KSlZxC01r5cVf8yye8mOS3Jr7XWbl57ywAAAICNGfIVg7TW3pjkjWtuy6liE1+jUMfeKF8de6uO/dAHdeytOvZDH9Sxd8pXx96qYz/0QR17q4790Ad17J3y96ylFykEAAAA9r8h1yAAAAAA9rvWmp8TfpI8I0lL8tAk70pyQ5JPJPmz/vf3J/mL/vc/TfLp/vcbktxtSdn3ntn2+HO/0t++Jcnnk3ysv/3m/jmP6tvzpAFt367847fPTvJ/k/zzbZ53IN2f8rh87Gt0wv1z25rk4Lz6V1FHkgcleUOSjya5Pslbkzx2TPlJzk9ybZIPJ3lvkmuSPGLm8WcmeV+Sm5PcmORlSe61oI6zk7w6yW19m96Z5BmL6poz525IsjWynptn5tVfzZTT+n8/kuR/z9z/mBXN2zaz7RfT/cnTG5K8alVzN8mVc/p28Qr27xuSbGXkPjF1/0tynySvTTdvb0n3da4Hj903+vn0tv71/kA/N+/ZP/bkJO9Jcmv/2IuXlL9tm/qfN/Zz59Ykv5nk+6eM95x63prkS/3zdzSnRs7ZNye5a5Ir0u2PNyW5LsmTJ471C/rX+aZ0x4kf6p8zqo4JY3H2srmwqjr6572k7/ddpvShf+xfJ/k/Sb7mhOdc2L8+H+hf099IcmhJPWOOgxf3z/ntJO8csY+PfV9a+BquYv8e248p47Go7yvox3NmxuVv0q2xbkhyxQr3vQNJ/kO6fe/4ti+cuO8tXWuMHI+564QT97N0a4Tj7T9pvTh0PLL4ve+H+tfv5r7fz19S7qjxSHJ1Rq5B5ozHvPeLVR3Pl63Nx645l60PTpoDGbnWGVDHdmuQhfNv5L59beasn/ttbkzymonHqbnjPWU8JtYx+PPYqf6z6w3Yiz/p3sDfnuRnZu67NMlLt9n2Z7Lk4LmgnpOem+QVJ+74SX6xb88rdlJ+kp/oy7l2m22fkuQP+52kprxGy9q6qP6d1pHk9CQfSvK0mfvOS3Lp0PL7g+SxzLxJJfnWJE/vf7+wP+ic098+LcmPJHnInPIr3YH22TP3nZvkucvqWjTnxtTT/76V5KZtnve4JG9Y17zt7782yeGd1jFv7szr29h5td1rPXafmNKHOWP3yCTfNqYP/Xz6eJJvmSn34v7+8/o+PLR/7ECSnxg5nx6Z5NvSvfF/18z9j09y3pTxXtb3Vc6pIXM23WLylUnuPvOaft/Y8pM8O91Ffc/sb39NkmdNqWPKWCyaC6sc73QfVj6R5I+SPG4H43xdP5cvnXn8vL7uh83c97QsXnyPPg4muVeST6b74P7AgWM99n1p4T4zpI5lYzqmH1PGY1nfV9WP/r5jSc4aUv6Efe8VSU7vb5+xqB+LXqcsWWusYDxm5+3c/SxzjpFj5m5OeO9LFya/N8n9Zubwj696PPrbj8vANciAeXuH1yIrOp7Pe60zYc254HVaOAf621sZv9a5Qz8yfA1yh7qHzKUMW9M+LF349+kkX7Xi8Z7yGWD0GiQTP4+dij+73oC99pPkq/vJ++AkH5i5/9LsQkDQT+Dbknxdks+kf3ObUn4/qc9Pl9qec8K2v57k+9KlZ98y8TVa2NZF9e+0jiQ/muSVOxnnJD+f5GcXPOftSR4/oo5/kuQP5jy2sK5Fc25MPf3jWzn1A4Jt5868vk2YVye91mP2ial9SPKEJG8bWe52c/fnkvzcnO1fleRHRpS/bZvShWHLzgIZPN7L+r7KObVszia5Z5I/T7+Q3clYp1vMf90224yuY8pYLJoLqxzvdMedNyZ5VpJfmTLO6Y7jtyT59iS/O3P/ryf54ZFjMPo4mO594z8n+ekkPzWgjrHvS0v3mSF1LBvTMf2YMh6L+r7KfvTbHMv0gGDZvnfGCvaLpWuNkeOxbN7O3c8yIiBYMHcvzR0DgrclecKYOTt2PGa2fVyGBwSD3y+ywuP5vNd6yDwY8TotnAP9NlvZeUCw3Rpkad1D5lKGrWl/Pt0ZJS9PcmRV4z11PCbUMfnz2Kn44xoEJ3t6kje11j6U5PNV9ehdbs8FST7WWvtouoXxU6YUUlUPSHKf1tp1+bvTgo8/do90B4k3JHlNkiNLipv3Gs1t66L6V1THP0yXeg+1XfnLyhhbx6Ltx5Y1tZ5T3oS5s8ig/XvCPrHQgj6cl+6slDG268OicsbWMW/7KW2dUs9u+Pokn2it3b6TQqrqjHQfQj66ojqmjMWmxvtIun3j6iRPraq7TmjP8TLenuQhVfX3+/unHNOmPOd4/UP38bHvS5vYv8f2Y8p4zGvXqvsx2cB97wsjipzX1iHzbMx4LCtv6H62zNDxW8kYLRmPKca0ayXH8yVOqTXcgjXIlLqnrJ/T1/kb2flxajtT+jG2jpV8HjtVCAhOdiTd91HS/7ujDwYrsKr2XJLuoLBdOU9N8tbW2peSvC7JM6rqtAltWtTWRfWvqo6/VVVXV9VNVfX6keXPlvGuqrq1ql6yzWOPqKobquqjVTXoA2tVXVlVN1bVu8fUNdaiek5RY+fOIkP3p7H7xDK70Qc2r9J9P3Ffq6q7pVsc/Va/CH9XkidOKOqSJK9trf2/JK9P8r3b1HXv/lj7oap6/og2LjwOVtXZ6T5IvKNf6H65qs5bUuyO3pcGGlXWxH7Ms2g8xvZx08epwfteVf1wP6c+2X9wml7pCWuNnY7H7Lxd4X6W7OHx2A8GrDnHlLWONdygNcjAukevn6vqm5L8WWvt40n+R5JHV9XXTuvKcqscjxl3qrXXgd1uwF5SVfdOd8rJeVXV0n3HvFXVC3apPacl+Z4kT6uqF6Y74N67qs4YmYQn3UQ+u6r+WX/7flX1oNbah/vHLqiqY/1j9073Pck3b9Omea/RZUvauqj+HdeR7mI6jz1eRmvtGVV1OMmLh5af7vtqj053caG01r65qi5O92ExfR2PTvfB8f1JHllVL01yj21f8W7775lp03Oq6qx0F4t705K6xlhUz34weO4sMnL/HrxPDLRtH9KN3cVDC1kyd78x/Xw6wc39YzcOrGZem25Od+rxqozq+5p9JMmhicfWv9Vau72q/rKq/kFr7bYV1DFlLI6P93ZzYVV1XJju+8Tvr6qkO6X3S+kuSjWo/Kr6hnQXlfr9voy7pTt988r83bH2xtban6c71j4/3amti/ox5jj4/Um+NsnH+vrPTLeIftF2G+/gfWnwPjNx/x7Vj0wYj0XHztafe7uCfuzImH2vtfbyJC+vqpv6tm1n0X6xaK0xZTzmzdsx+9lcI9/7jh8/3jKmjhMtGY8pxrxfrOR4PqA9g9acA8ta9xpu0RpkcN07WD8fSfLQmTXVmX29L5vT3rHrgynjMbiOFX8eOyU4g+COLk73fcFzW2tbrbUHpLuC5bfuUnu+I90C6QF9e85N97+ZTx9TSFU9JN0FQc7py9lKcnmSS6rqzHT9OzTz2HMyPxmb9xq9aF5bF9W/qjrSXYH1gqp62kw59xxZ/u8lubSqHjOnjMuTvLiq7j9z37xwIOneYE+vqn+xTXlXLqlrjEX1nNImzJ1FBu3fE/aJnfThLUnuXlU/PrP9N1XVvA8V8/rw5iTPqqpvninnmVV1nyT/Mcm/r6oH9/ffpar+zYImb9umdIuux1TVRTP3X1hVjxjxciytZ0Hf16Y/U+RXk/xydf9rl6q6b1U9c0Jxl6f7UHVmX86ZVXV0Yh1TxuKlmT8XVlXHkSQ/NjOfH5jkiVW13XFnXvkvSXeBq63+535Jzqmqc9NdCOqFVfWwmXKWHdPGHgePJLlwpg/fmMXHlanvS2P2mSn799h+TBmPsWujKf1YhWX73kur6vT+sdPShSDzLNovFq01pozHvHk7Zj9bZMz4XZ7kF4+PSVXdvar+1cj6Zss6aTwmljX4/WLFx/N5xqw5l1nrGm7AGmRM3aPXz1V1l3RnI33DTP3fncVrqrHrgynjMaaOlXweO5UICO7oSLrvec16XZIf2IW2JKtrz7xyjiT5p0ne0lr765nHfjtdSnb3EWX94wVtXVT/mPbOraO19lfpkspnV9VtVfXOdAu3XxhR/g+kS/8vr6qPVNX/THcwfGmStNbemOSXk/xOVd3SP/6VdFfpPUn/PytPT/LtVfWxqrouXcr671prf7qorjEW1TO2rD1o7NyZUtaJ+9PYfWJqvUf6sXtGku+s7usqN6e7sNBnRpZ1Sf/z4qr6YFXdmu5q27e31t6X5CeTvKa//6Yk953X2CVtemqS51bVh6vqlnQXuPrs8pdgdD274UXp/tzXLdX9z+Jv9bfH+i/pLmz57r6cP0j3v36j65gyFq21/5U5c2FFddye5EmZ+V/M1tpfJnlHku8aUf7jcvJcvjrJJa07Q+t5SV5VVR+oqj9MdwXsVy95rQYdB6tqK8mhdFeGP/78jyW5ffbD6wl28r40dJ8Zu3//vbH9mDIeC9o1by0y+jg1p5yxFu17L0zyJ0luqqo/TnedhVdmzvFmwH5x0lpjyrxaMG9/OiP2syUGj1+/zrkyyZv7Pl+f6WcbLxqPUSa8X6zqeD6vPWPWnMvKWvcabtkaZEzdU9bPj03y6dbap2ee87YkD6+qbdciY8d7yniMrGOvfT5cu2onnx0GAAAA3Mk4gwAAAAAQEAAAAAACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACACAgAAACDJ/wcXVvxd3jp10wAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1296x432 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot of Kmers Freq\n",
    "plt.figure(figsize=(18,6))\n",
    "plt.bar(dna1_kfreq.keys(),dna1_kfreq.values())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Checking Distance Between Sequence / Similarity\n",
    "+ Hamming distance\n",
    " - Must have the same length\n",
    "+ Check dissimilarity\n",
    "![](./images/aligngap.png)\n",
    "+ For use in sequence clustering, phylogenetic reconstruction\n",
    "+ Mismatch and Match Frequency\n",
    "+ Kmer-Distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq1 = DNA('ACTATCAGT')\n",
    "seq2 = DNA('ACTTACAGT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate the matches/mismatches frequency\n",
    "seq1.match_frequency(seq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True,  True,  True, False, False,  True,  True,  True,  True])"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate the matches/mismatches as boolean\n",
    "seq1.matches(seq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate the matches/mismatches frequency\n",
    "seq1.mismatch_frequency(seq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([False, False, False,  True,  True, False, False, False, False])"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate the matches/mismatches as boolean\n",
    "seq1.mismatches(seq2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Hamming distance: shows how many places 2 strings differ\n",
    "\n",
    "+ Hamming distance between two strings of equal length is the number of positions at which the corresponding symbols are different.\n",
    "+ In other words, it measures the minimum number of substitutions required to change one string into the other, or the minimum number of errors that could have transformed one string into the othe\n",
    "+ It is used for error detection or error correction\n",
    "+ It is used to quantify the similarity of DNA sequences,\n",
    "+ For checking the edit distance\n",
    "        edit distance is a way of quantifying how dissimilar two strings (e.g., words) are to one another by counting the minimum number of operations required to transform one string into the other.\n",
    "        eg Levenshtein distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.2222222222222222"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# hamming distance\n",
    "seq1.distance(seq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Custom Hamming Distance Fxn\n",
    "def hamming_distance(lhs,rhs):\n",
    "    return len([(x,y) for x,y in zip(lhs,rhs) if x != y])\n",
    "\n",
    "hamming_distance(seq1,seq2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Kmer Distance\n",
    "+ The kmer distance between two sequences is the fraction of kmers that are unique to either sequence.\n",
    "+ Find kmers that are not found in each sequence orare dissimilar in each aka the sum of the non-intersection\n",
    "+ Using set ^ set for dissimilar/symmetric_difference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACT': 1, 'CTA': 1, 'TAT': 1, 'ATC': 1, 'TCA': 1, 'CAG': 1, 'AGT': 1}"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Kmer distance\n",
    "from skbio.sequence.distance import kmer_distance\n",
    "\n",
    "# Kmers freq in Seq\n",
    "seq1.kmer_frequencies(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACT': 1, 'CTT': 1, 'TTA': 1, 'TAC': 1, 'ACA': 1, 'CAG': 1, 'AGT': 1}"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Kmers freq in Seq\n",
    "seq2.kmer_frequencies(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.7272727272727273"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the kmer distance\n",
    "kmer_distance(seq1,seq2,k=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Custom Fxn For Kmer Distance\n",
    "+ Set\n",
    "+ Intersection\n",
    "+ Dissimilar Set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Method 1\n",
    "def custom_kmer_distance(seq1,seq2):\n",
    "    seq1_set = set(seq1)\n",
    "    seq2_set = set(seq2)\n",
    "    union_seq = seq1_set.union(seq2_set)\n",
    "    intersection_seq = seq1_set.intersection(seq2_set)\n",
    "    disimilarity = len(union_seq) - len(intersection_seq)\n",
    "    print(disimilarity)\n",
    "    distance = disimilarity/len(union_seq)\n",
    "    return distance\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Kmers freq in our seq\n",
    "seq1_k = seq1.kmer_frequencies(3).keys()\n",
    "\n",
    "# Kmers freq in our seq\n",
    "seq2_k = seq2.kmer_frequencies(3).keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.7272727272727273"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "custom_kmer_distance(seq1_k,seq2_k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Method 2\n",
    "def custom_kmer_distance2(seq1,seq2):\n",
    "    seq1_set = set(seq1)\n",
    "    seq2_set = set(seq2)\n",
    "    union_seq = seq1_set.union(seq2_set)\n",
    "    disimilarity = seq1_set ^ seq2_set\n",
    "    print(disimilarity)\n",
    "    distance = len(disimilarity)/len(union_seq)\n",
    "    return distance\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'CTT', 'TAT', 'ACA', 'TCA', 'CTA', 'ATC', 'TTA', 'TAC'}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.7272727272727273"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "custom_kmer_distance2(seq1_k,seq2_k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACA', 'ATC', 'CTA', 'CTT', 'TAC', 'TAT', 'TCA', 'TTA'}"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Symmetric Difference\n",
    "set(seq1_k).symmetric_difference(set(seq2_k))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequence Alignment\n",
    "+ Global : Needle\n",
    "+ Local : Smith-Waterman\n",
    "+ Terms\n",
    "    - Matches\n",
    "    + Mismatch\n",
    "    + Gaps\n",
    "\n",
    "![](./images/aligngap.png)\n",
    "\n",
    "![](./images/localvsglobalalignment.jpg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Alignment\n",
    "import skbio.alignment as align\n",
    "\n",
    "# dir(align)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACTATCAGT\n",
      "ACTTACAGT\n"
     ]
    }
   ],
   "source": [
    "# Local Alignment\n",
    "print(seq1)\n",
    "print(seq2)\n",
    "seq3 = DNA('ACTTAC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Local Alignment StripedSmithWaterman\n",
    "alignment,score,start_end_positions = align.local_pairwise_align_ssw(seq1,seq2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TabularMSA[DNA]\n",
       "---------------------\n",
       "Stats:\n",
       "    sequence count: 2\n",
       "    position count: 4\n",
       "---------------------\n",
       "CAGT\n",
       "CAGT"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the alignment\n",
    "alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(5, 8), (5, 8)]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get Postions\n",
    "start_end_positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GLobal Aligment StripedSmithWaterman\n",
    "glb_alignment,glb_score,glb_start_end_positions = align.global_pairwise_align_nucleotide(seq1,seq2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TabularMSA[DNA]\n",
       "---------------------\n",
       "Stats:\n",
       "    sequence count: 2\n",
       "    position count: 9\n",
       "---------------------\n",
       "ACTATCAGT\n",
       "ACTTACAGT"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glb_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.0"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glb_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dissimilar seq length\n",
    "# GLobal Aligment StripedSmithWaterman\n",
    "glb_alignment,glb_score,glb_start_end_positions = align.global_pairwise_align_nucleotide(seq1,seq3)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TabularMSA[DNA]\n",
       "----------------------\n",
       "Stats:\n",
       "    sequence count: 2\n",
       "    position count: 13\n",
       "----------------------\n",
       "----ACTATCAGT\n",
       "ACTTAC-------"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glb_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.0"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glb_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Comparing Protein Sequence Alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "from skbio import Protein\n",
    "\n",
    "p1 = Protein('MITITV')\n",
    "p2 = Protein('ITITV')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "local_alignment_protein  = align.local_pairwise_align_protein(p1,p2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(TabularMSA[Protein]\n",
       " ---------------------\n",
       " Stats:\n",
       "     sequence count: 2\n",
       "     position count: 5\n",
       " ---------------------\n",
       " ITITV\n",
       " ITITV,\n",
       " 25.0,\n",
       " [(1, 5), (0, 4)])"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "local_alignment_protein "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phylogenetics\n",
    "+ Phylogenetics: study of evolutionary history and relationships among entities\n",
    "+ A Phylo Tree/Evolutionary Tree is a branching diagram or tree showing the evolutionary relationships among various\n",
    "species or entities based on similarity and differences in their physical or genetic characteristics\n",
    "\n",
    "+ Newick Format\n",
    "+ Nexus\n",
    "+ PhyloXML\n",
    "\n",
    "#### Terms\n",
    "+ (;)The semi-colon creates a root node to its left. \n",
    "    - Recall that there can only be one root in a tree.\n",
    "+ (,)The comma which creates a node to its right. \n",
    "+  (( )) The paired parenthesis provide  relationships,\n",
    "+ Placing ( ) in a nodes location will create a child inside the parenthesis on the left-most inner edge.\n",
    "+ children <= node <= root\n",
    "\n",
    "+ \"((grand_child, grand_child)Brother, Children)rootParent;\"\n",
    "+ \"((Enoch,Awan)Seth,Cain,Abel)Adam;\"\n",
    "\n",
    "![](./images/noahfamilytree.jpg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Terms\n",
    "+ There are several symbols which define nodes, the first of which is the semi-colon (;). The semi-colon creates a root node to its left. Recall that there can only be one root in a tree.\n",
    "\n",
    "The next symbol is the comma (,), which creates a node to its right. However, these two alone are not enough. For example imagine the following string: , , , ;. It is evident that there is a root, but the other 3 nodes, defined by commas, have no relationship. For this reason, it is not a valid Newick string to have more than one node at the root level.\n",
    "\n",
    "To provide these relationships, there is another structure: paired parenthesis (( )). These are inserted at the location of an existing node and give it the ability to have children. Placing ( ) in a nodes location will create a child inside the parenthesis on the left-most inner edge."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    ">>> from io import StringIO\n",
    ">>> from skbio import read\n",
    ">>> from skbio.tree import TreeNode\n",
    ">>> f = StringIO(\"((D, E)B, (F, G)C)A;\")\n",
    ">>> tree = read(f, format=\"newick\", into=TreeNode)\n",
    ">>> f.close()\n",
    ">>> print(tree.ascii_art())\n",
    "                    /-D\n",
    "          /B-------|\n",
    "         |          \\-E\n",
    "-A-------|\n",
    "         |          /-F\n",
    "          \\C-------|\n",
    "        \n",
    "### Method 2\n",
    ">>> from io import StringIO\n",
    ">>> from skbio import read\n",
    ">>> from skbio.tree import TreeNode\n",
    ">>> f = StringIO(\"((D, E)B, (F, G)C)A;\")\n",
    ">>> tree = read(f, format=\"newick\", into=TreeNode)\n",
    ">>> f.close()\n",
    ">>> print(tree.ascii_art())\n",
    "                    /-D\n",
    "          /B-------|\n",
    "         |          \\-E\n",
    "-A-------|\n",
    "         |          /-F\n",
    "          \\C-------|\n",
    "                    \\-G                    \\-G\n",
    "            \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'((grand_child, grand_child)Brother, Child)rootParent;'"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"((grand_child, grand_child)Brother, Child)rootParent;\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from skbio import TreeNode\n",
    "from io import StringIO\n",
    "\n",
    "# Example 2\n",
    "# Create A TreeNode\n",
    "noah_tree = TreeNode.read(StringIO(\"((Magog,Madia,Javan, Gomer)Japheth,(Israel,Arphaxad)Shem,(Cush,Egypt,Phut,Canaan)Ham)root;\"))\n",
    "\n",
    "# Check if this is a root\n",
    "noah_tree.is_root()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check for if tip\n",
    "noah_tree.is_tip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                    /-Magog\n",
      "                   |\n",
      "                   |--Madia\n",
      "          /Japheth-|\n",
      "         |         |--Javan\n",
      "         |         |\n",
      "         |          \\-Gomer\n",
      "         |\n",
      "         |          /-Israel\n",
      "-root----|-Shem----|\n",
      "         |          \\-Arphaxad\n",
      "         |\n",
      "         |          /-Cush\n",
      "         |         |\n",
      "         |         |--Egypt\n",
      "          \\Ham-----|\n",
      "                   |--Phut\n",
      "                   |\n",
      "                    \\-Canaan\n"
     ]
    }
   ],
   "source": [
    "# In A Dendogam\n",
    "print(noah_tree.ascii_art())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![](./images/noahfamilytree.jpg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dir(noah_tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<TreeNode, name: Japheth, internal node count: 0, tips count: 4>,\n",
       " <TreeNode, name: Shem, internal node count: 0, tips count: 2>,\n",
       " <TreeNode, name: Ham, internal node count: 0, tips count: 4>]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "noah_tree.children"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "noah_tree.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method TreeNode.lowest_common_ancestor of <TreeNode, name: root, internal node count: 3, tips count: 10>>"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "noah_tree.lowest_common_ancestor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'(((Methu)Enoch,Awan)Seth,Cain,Abel)Adam;'"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example 2\n",
    "\"(((Methu)Enoch,Awan)Seth,Cain,Abel)Adam;\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example 2\n",
    "# Create A TreeNode\n",
    "adam_tree = TreeNode.read(StringIO(\"(((Methu)Enoch,Awan)Seth,Cain,Abel)Adam;\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                    /Enoch--- /-Methu\n",
      "          /Seth----|\n",
      "         |          \\-Awan\n",
      "-Adam----|\n",
      "         |--Cain\n",
      "         |\n",
      "          \\-Abel\n"
     ]
    }
   ],
   "source": [
    "print(adam_tree.ascii_art())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading Files\n",
    "+ FASTA\n",
    "+ GENBANK\n",
    "+ NEWICK\n",
    "+etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<generator object IORegistry.read.<locals>.<genexpr> at 0x7f8af41cedd0>\n"
     ]
    }
   ],
   "source": [
    "# Load pkg\n",
    "import skbio.io\n",
    "\n",
    "# Reading A FASTA file\n",
    "# Method 1\n",
    "seq_file = skbio.io.read(\"./datasets/sequence.fasta\", format='fasta')\n",
    "\n",
    "\n",
    "print(seq_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for i in seq_file:\n",
    "#     print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "from skbio import DNA,Sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sequence\n",
       "-----------------------------------------------------------------------\n",
       "Metadata:\n",
       "    'description': 'Severe acute respiratory syndrome coronavirus 2\n",
       "                    isolate Wuhan-Hu-1, complete genome'\n",
       "    'id': 'MN908947.3'\n",
       "Stats:\n",
       "    length: 29903\n",
       "-----------------------------------------------------------------------\n",
       "0     ATTAAAGGTT TATACCTTCC CAGGTAACAA ACCAACCAAC TTTCGATCTC TTGTAGATCT\n",
       "60    GTTCTCTAAA CGAACTTTAA AATCTGTGTG GCTGTCACTC GGCTGCATGC TTAGTGCACT\n",
       "...\n",
       "29820 TTTAGTAGTG CTATCCCCAT GTGATTTTAA TAGCTTCTTA GGAGAATGAC AAAAAAAAAA\n",
       "29880 AAAAAAAAAA AAAAAAAAAA AAA"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Method 2\n",
    "# Directly into A Sequence\n",
    "seq_file2 = Sequence.read('./datasets/sequence.fasta')\n",
    "\n",
    "seq_file2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dir(seq_file2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DNA\n",
       "-----------------------------------------------------------------------\n",
       "Metadata:\n",
       "    'ACCESSION': 'MN908947'\n",
       "    'COMMENT': <class 'str'>\n",
       "    'DEFINITION': 'Severe acute respiratory syndrome coronavirus 2\n",
       "                   isolate Wuhan-Hu-1, complete genome.'\n",
       "    'KEYWORDS': '.'\n",
       "    'LOCUS': <class 'dict'>\n",
       "    'REFERENCE': <class 'list'>\n",
       "    'SOURCE': <class 'dict'>\n",
       "    'VERSION': 'MN908947.3'\n",
       "Interval metadata:\n",
       "    23 interval features\n",
       "Stats:\n",
       "    length: 29903\n",
       "    has gaps: False\n",
       "    has degenerates: False\n",
       "    has definites: True\n",
       "    GC-content: 37.97%\n",
       "-----------------------------------------------------------------------\n",
       "0     ATTAAAGGTT TATACCTTCC CAGGTAACAA ACCAACCAAC TTTCGATCTC TTGTAGATCT\n",
       "60    GTTCTCTAAA CGAACTTTAA AATCTGTGTG GCTGTCACTC GGCTGCATGC TTAGTGCACT\n",
       "...\n",
       "29820 TTTAGTAGTG CTATCCCCAT GTGATTTTAA TAGCTTCTTA GGAGAATGAC AAAAAAAAAA\n",
       "29880 AAAAAAAAAA AAAAAAAAAA AAA"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Method 3\n",
    "# Directly into A Sequence\n",
    "seq_file3 = DNA.read('./datasets/sequence.fasta')\n",
    "\n",
    "# Method 3\n",
    "# Directly into A Sequence\n",
    "seq_file4 = DNA.read('./datasets/sequence.gb')\n",
    "\n",
    "seq_file4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Writing Sequences\n",
    "```python\n",
    "with StringIO() as fh:\n",
    "    print(dna_seq.write(fh).getvalue())\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Quering NCBI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GeneticCode (Trematode Mitochondrial)\n",
       "-------------------------------------------------------------------------\n",
       "  AAs  = FFLLSSSSYY**CCWWLLLLPPPPHHQQRRRRIIMMTTTTNNNKSSSSVVVVAAAADDEEGGGG\n",
       "Starts = -----------------------------------M---------------M------------\n",
       "Base1  = UUUUUUUUUUUUUUUUCCCCCCCCCCCCCCCCAAAAAAAAAAAAAAAAGGGGGGGGGGGGGGGG\n",
       "Base2  = UUUUCCCCAAAAGGGGUUUUCCCCAAAAGGGGUUUUCCCCAAAAGGGGUUUUCCCCAAAAGGGG\n",
       "Base3  = UCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAG"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_mit = skbio.sequence.GeneticCode.from_ncbi(21)\n",
    "species_mit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dir(species_mit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Trematode Mitochondrial'"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_mit.name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 2, 3, -1, -2, -3]"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_mit.reading_frames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TIS\n",
      "LS\n",
      "YQ\n",
      "TDS\n",
      "LM\n",
      "W*\n"
     ]
    }
   ],
   "source": [
    "mRNA1 = seq1.transcribe()\n",
    "\n",
    "for i in species_mit.translate_six_frames(mRNA1):\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
